src/sampler.jl

```diff
@@ -50,28 +102,25 @@ function sample(
     Hs = Vector{T}(undef, n_samples)
     αs = Vector{T}(undef, n_samples)
     stats = Vector{NamedTuple}(undef, n_samples)
-    # Prepare phase point for sampling
-    h = update(h, θ) # Ensure h.metric has the same dim as θ.
-    r = rand(rng, h.metric)
-    z = phasepoint(h, θ, r)
+    h, z = init(rng, h, θ)
+    # Progress meter
     pm = progress ? Progress(n_samples, desc="Sampling", barlen=31) : nothing
     time = @elapsed for i = 1:n_samples
-        z, stat = transition(rng, τ, h, z)
-        θs[i], αs[i], Hs[i], stats[i] = z.θ, stat.acceptance_rate, stat.hamiltonian_energy, stat
-        showvalues = Dict{Symbol,Any}(pairs(stat)...)
-        if !(adaptor isa Adaptation.NoAdaptation)
-            if i <= n_adapts
-                adapt!(adaptor, θs[i], αs[i])
-                i == n_adapts && finalize!(adaptor)
-                h, τ = update(h, τ, adaptor)
-                (i == n_adapts && verbose && !progress) && @info "Finished $n_adapts adapation steps" adaptor τ.integrator h.metric
-            end
-            # Progress info for adapation
-            progress && (showvalues[:step_size] = τ.integrator.ϵ; showvalues[:precondition] = h.metric)
+        z, θs[i], αs[i], Hs[i], stats[i] = step(rng, h, τ, z)
+        # Adapt h and τ; what mutable is the adaptor
+        h, τ, isadapted = adapt!(h, τ, adaptor, i, n_adapts, θs[i], αs[i])
+        # Adapation finish
+        if isadapted && i == n_adapts && verbose && !progress
+            @info "Finished $n_adapts adapation steps" adaptor τ.integrator h.metric
+        end
+        if progress
```

Comment:
Perhaps wrap this block of code into a re-usable function, e.g.

```julia
ProgressMeter.next!(pm; v::NamedTuple) == ProgressMeter.next!(pm; showvalues=collect(zip(keys(v), values(v))))
```

then write

```julia
v=(mass_matrix=h.metric, iteration=i, stats...)
progress && ProgressMeter.next!(pm; showvalues=v)
```

Reply:
Good idea!

Reply:
Done in 964757d.